src/submit_jobs.py

```diff
@@ -0,0 +1,116 @@
+DEFAULT_VAR_USE_PARTITIONS = True
+
+# Constants for paths inside the Spark container
+PATH_IPTS_PASSWORD = "/run/secrets/IPTS_PASSWORD"
+PATH_PREDICATES = "/tmp/config/default_predicates.csv"
+PATH_INITIAL_DIR = "/tmp/target/initial"
+PATH_FINAL_DIR = "/tmp/target/final"
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Submit iasWorld Spark extraction jobs"
+    )
+    parser.add_argument(
+        "--json-file",
+        type=str,
+        nargs="?",
+        help="Path to a JSON file containing job configuration(s)",
+    )
+    parser.add_argument(
+        "--json-string",
+        type=str,
+        nargs="?",
+        help="JSON string containing job configurations(s)",
+    )
+
+    return parser.parse_args()
+
+
+def main() -> str:
+    # Get the job definition(s) from the argument JSON
+    args = parse_args()
+    job_config = load_job_config(args)
+
+    # Predicates are used to slice up a table/job into discrete chunks to
+    # improve parallelization. In our case, we slice tables into equally-sized
+    # chunks by PARID (where available) since it is indexed (we get fast reads)
+    predicates_csv = read_predicates(PATH_PREDICATES)
+
+    # Each session is shared across all read jobs and manages job order,
+    # credentialing, retries, etc.
+    current_datetime = datetime.now().strftime("%Y-%m-%d_%H:%M:%S")
+    session_name = f"iasworld_{current_datetime}"
+    session = SharedSparkSession(
+        app_name=session_name, password_file_path=PATH_IPTS_PASSWORD
+    )
+
+    # For each Spark job, get the table structure based on the job definition
+    # or defaults. Then, perform the JDBC read of the job to extract data.
+    # Finally, append the job to a list so we can use it later
+    jobs = []
+    for job_name, config in job_config.items():
+        use_partitions = config.get(
+            "use_partitions", DEFAULT_VAR_USE_PARTITIONS
+        )
+        if use_partitions:
+            min_year = config.get("min_year", DEFAULT_VAR_MIN_YEAR)
+            max_year = config.get("max_year", DEFAULT_VAR_MAX_YEAR)
+            cur = config.get("cur", DEFAULT_VAR_CUR)
+            years = [x for x in range(min_year, max_year + 1)]
+        else:
+            years, cur = None, None
+
+        if config.get("use_predicates", DEFAULT_VAR_USE_PREDICATES):
+            predicates = construct_predicates(predicates_csv, years)
+        else:
+            predicates = None
+
+        spark_job = SparkJob(
+            session=session,
+            table_name=config.get("table_name"),
+            taxyr=years,
+            cur=cur,
+            predicates=predicates,
+            use_partitions=use_partitions,
+            initial_dir=PATH_INITIAL_DIR,
+            final_dir=PATH_FINAL_DIR,
+        )
+
+        # Run the Spark read job. Each job will create n_predicates X n_years
+        # files in the target/initial/ dir, assuming predicates are enabled
+        spark_job.read()
```

Comment:
**[Question, non-blocking]** If I'm understanding correctly, this blocks until the full matrix of predicate reads is complete, is that right? If so, it's making me wonder if there's possibly even more performance that we could eke out by parallelizing the jobs themselves too, such that workers within a job could move onto the next job if all predicate reads have been picked up but not yet completed. Highly speculative, so not something I expect us to investigate now!

Reply:
Yep that's right. That's a good idea! Added in fa34a59. It doesn't seem to have much impact on performance, but it also doesn't seem to hurt so 🤷 .